File: benchmarks/semantic_search/fixtures/typescript-workflow-discovery-v1/selection/retry.ts
import type { WorkflowValidator } from "./validator.js";

export function selectWithRetry(
  validator: WorkflowValidator,
  workflowId: string,
  correctedId: string,
): string {
  try {
    validator.validate(workflowId);
    return workflowId;
  } catch {
    return selfCorrectSelection(validator, correctedId);
  }
}

export function selfCorrectSelection(validator: WorkflowValidator, workflowId: string): string {
  validator.validate(workflowId);
  return workflowId;
}

File: benchmarks/semantic_search/fixtures/rust-workflow-discovery-v1/src/selection/retry.rs
use super::validator::{ValidationError, WorkflowValidator};

pub fn select_with_retry(
    validator: &WorkflowValidator,
    workflow_id: &str,
    corrected_id: &str,
) -> Result<String, ValidationError> {
    match validator.validate(workflow_id) {
        Ok(()) => Ok(workflow_id.to_owned()),
        Err(_) => self_correct_selection(validator, corrected_id),
    }
}

pub fn self_correct_selection(
    validator: &WorkflowValidator,
    workflow_id: &str,
) -> Result<String, ValidationError> {
    validator.validate(workflow_id)?;
    Ok(workflow_id.to_owned())
}
